### PR TITLE
fix(api): harden abuse-gate error logging

### PR DESCRIPTION
**Pull request by DysektAI**
## Summary

- redact `CF-Connecting-IP` from abuse-gate binding error names/messages before structured logging
- remove obsolete rate-limiter `503` responses from the OpenAPI contract now that the daily quota deliberately fails open
- correct legacy Durable Object retention documentation and track staged cleanup in #594

Follow-up to #591.

## Production remediation

- provisioned the missing `IP_HASH_SECRET` on the production `api-gateway` Worker
- verified the production health and OpenAPI endpoints

## Validation

- `pnpm run typecheck`
- `pnpm run lint`
- `pnpm run test` (218 files, 2160 tests)
- `pnpm run test:api-gateway` (7 files, 106 tests)
- `pnpm run i18n:check`
- `pnpm run validate:openapi`
- `pnpm --filter api-gateway run typecheck`
- `pnpm --filter api-gateway exec wrangler deploy --dry-run --config wrangler.toml`
- targeted Prettier check

---

<a href="https://app.devin.ai/review/tarkovtracker-org/tarkovtracker/pull/595" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

<a href="https://cubic.dev/pr/tarkovtracker-org/TarkovTracker/pull/595?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Introduced tiered daily read and write quotas, resetting at 00:00 UTC.
  * Added a pre-authentication IP abuse gate to help protect token validation from excessive traffic.
  * Added clearer rate-limit response headers and retry guidance.

* **Documentation**
  * Updated API, operational, and OpenAPI documentation to describe the new quota and abuse-gate behavior.
  * Clarified fail-open handling and related monitoring guidance.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR hardens the abuse-gate error-logging path in the Cloudflare API Gateway Worker by redacting raw `CF-Connecting-IP` values from structured log output before they reach Worker logs, and by cleaning up the OpenAPI contract and documentation to match the current fail-open daily-quota behavior.

- **IP redaction in abuse-gate logging**: Adds `sanitizeLogValue` and `errorInfo` helpers that run `String.replaceAll` on the raw IP before any `console.warn` call, replacing it with `[redacted]`. A lazy HMAC-SHA-256 hash (`getIpHash`) is logged instead, preserving debuggability without PII exposure.
- **OpenAPI contract cleanup**: Removes the now-obsolete `503` responses that were added when the rate limiter could hard-fail; they no longer match the fail-open behavior and are removed from all endpoint response tables.
- **Limits / docs cleanup**: Drops the per-tier burst column (replaced by the infrastructure-level abuse gate) from `docs/API.md` and removes the matching `burstPerMinute` field from `limits.ts` to keep code and docs in sync.
</details>

<details open><summary><h3>Confidence Score: 4/5</h3></summary>

Safe to merge. The redaction helpers are correct, all abuse-gate logging paths are covered, and the new tests explicitly assert the raw IP never reaches log output.

The core IP-redaction logic in sanitizeLogValue/errorInfo is straightforward and the lazy ip-hash closure is correct. The test suite adds explicit assertions that the raw IP string is absent from warn output and that the hashed form matches an independently computed HMAC, giving strong confidence the main goal of the PR is achieved.

**Files Needing Attention:** No files require special attention; wrangler.toml carries a pre-existing namespace_id that cannot be validated from the diff but is not touched by this PR.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| workers/api-gateway/src/index.ts | Adds sanitizeLogValue/errorInfo helpers and lazy ip-hash computation; correctly redacts raw CF-Connecting-IP from all abuse-gate log paths before structured output. Logic and edge cases are sound. |
| workers/api-gateway/src/__tests__/rate-limits.test.ts | New tests cover abuse-gate fail-open, 429 logging with hashed IP, and explicit assertion that the raw IP string does not appear anywhere in the warn log; strong coverage of the new redaction path. |
| workers/api-gateway/src/limits.ts | Removed burstPerMinute from TierLimits and TIER_LIMITS; retained ABUSE_GATE_PERIOD_SEC and daily quota constants. Consistent with infrastructure-level abuse gate replacing per-tier burst enforcement. |
| workers/api-gateway/src/openapi.ts | Removed 503 response codes from all endpoint definitions; the RateLimited component response now correctly documents only 429 with dual descriptions for abuse-gate vs daily-quota scenarios. Aligns with fail-open behavior. |
| workers/api-gateway/src/types.ts | Added optional IP_HASH_SECRET to the Env interface; marked optional so local dev and tests without the binding continue to work. No other changes. |
| workers/api-gateway/wrangler.toml | Retains the [[ratelimits]] binding with namespace_id 1001 and period 60 (in sync with ABUSE_GATE_PERIOD_SEC). Migrations section corrected to use new_sqlite_classes only. Cannot verify production namespace_id is correct from diff alone. |
| docs/API.md | Removed Burst/min column from tier tables and updated the description to drop burst-limit references; consistent with the code changes in limits.ts. |
| docs/RATE_LIMITING.md | Updated to reflect that burst enforcement is now handled by the infrastructure abuse gate, not per-tier DO counters. Cleanup tracking noted for #594. |
| docs/runbook.md | Runbook updated with corrected Durable Object retention guidance and IP_HASH_SECRET provisioning steps. Documentation accurately reflects the hardened logging behavior. |

</details>

<details open><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant Client
    participant Worker as API Gateway Worker
    participant AbuseGate as API_ABUSE_LIMITER
    participant Hash as hashIp (HMAC-SHA-256)
    participant Log as Worker Logs
    participant DailyDO as ApiGatewayRateLimiter (DO)
    participant Supabase

    Client->>Worker: Request + CF-Connecting-IP + Bearer token
    Worker->>Worker: Extract clientIp

    alt clientIp present AND API_ABUSE_LIMITER bound
        Worker->>AbuseGate: "limit key=api:raw-ip"
        alt Binding throws
            AbuseGate-->>Worker: Error may contain raw IP
            Worker->>Hash: hashIp(clientIp, IP_HASH_SECRET)
            Hash-->>Worker: 16-hex HMAC digest
            Worker->>Worker: errorInfo replaces IP with redacted
            Worker->>Log: WARN abuse_gate_unavailable ip_hash + redacted error
            Note over Worker: Fail open
        else Blocked
            AbuseGate-->>Worker: success false
            Worker->>Hash: hashIp(clientIp)
            Worker->>Log: WARN abuse_gate_429 ip_hash
            Worker-->>Client: 429 Retry-After 60
        else Allowed
            AbuseGate-->>Worker: success true
        end
    end

    Worker->>Supabase: validateToken
    Worker->>DailyDO: POST daily-kind:user_id
    DailyDO-->>Worker: allowed remaining resetAt

    alt Quota unavailable
        Worker->>Log: WARN daily_quota_unavailable
        Note over Worker: Fail open
    else Quota exceeded
        Worker-->>Client: 429 X-RateLimit headers
    else Allowed
        Worker->>Supabase: Handle request
        Worker-->>Client: 200
    end
```
</details>

<sub>Reviews (1): Last reviewed commit: ["fix(api): harden abuse-gate error loggin..."](https://github.com/tarkovtracker-org/tarkovtracker/commit/1c01b671246fb4f4114d646b78de456da787c7ea) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=47122764)</sub>

**Context used:**

- Context used - AGENTS.md ([source](https://app.greptile.com/dysektai/github/tarkovtracker-org/tarkovtracker/-/custom-context?memory=635c24fb-71e5-43a4-9979-74405b725e95))

### CODE REVIEW

File: workers/api-gateway/src/index.ts [hunk 1]
```diff
@@ -371,87 +248,80 @@ async function rateLimit(
...
+      resetAt <= now ||
+      resetAt > now + windowSec * 1000
+    ) {
+      return {
+        allowed: false,
+        unavailable: true,
+        reason: 'bad_json',
+        message: 'Rate limiter unavailable',
+      };
+    }
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/index.ts:292]: Confirmed. This is the deliberate fail-open behavior documented in `docs/RATE_LIMITING.md` and covered by malformed/unavailable daily-quota tests. Legitimate deny payloads still pass validation and remain fail-closed at 429; no change needed.

File: workers/api-gateway/src/index.ts [hunk 2]
```diff
@@ -663,6 +530,53 @@ async function authenticateAndRateLimit(
...
+        JSON.stringify({
+          event: 'abuse_gate_429',
+          action,
+          ip_hash: await getIpHash(),
+        })
+      );
+      return errorResponse('Too many requests', 429, envOrigin, requestOrigin, {
+        'Retry-After': String(ABUSE_GATE_PERIOD_SEC),
+      });
+    }
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/src/index.ts:578]: Confirmed. The pre-auth gate is intentionally coarse infrastructure protection rather than a customer entitlement; `docs/API.md` and `docs/RATE_LIMITING.md` explicitly document its per-IP/shared-NAT behavior. No change needed.

File: workers/api-gateway/wrangler.toml [hunk 3]
```diff
@@ -38,6 +38,18 @@ bindings = [
...
+# it shields the api_tokens lookup from floods and is not a customer quota.
+# Counters are per-location and eventually consistent. `period` must be 10 or
+# 60 and must stay in sync with ABUSE_GATE_PERIOD_SEC in src/limits.ts.
+[[ratelimits]]
+name = "API_ABUSE_LIMITER"
+namespace_id = "1001"
+
+  [ratelimits.simple]
+  limit = 300
+  period = 60
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply workers/api-gateway/wrangler.toml:51]: Confirmed. The binding schema and period were validated with Wrangler dry-run, and namespace `1001` is unique across this repository’s Worker configurations. Account operators must preserve that namespace uniqueness for any separately managed Workers. No code change needed.

File: docs/runbook.md [hunk 4]
```diff
@@ -121,9 +121,10 @@ product.
    ```
 
    The api-gateway Worker auto-deploys from `main` on merge. If `IP_HASH_SECRET` is absent at
-   deploy time, every 429 and `ip_backstop_unavailable` log line emits `ip_hash: null`, defeating
-   the IP-level abuse observability the change introduced. Provision the secret **before** merging
-   so the first post-merge request already has a non-null HMAC identifier. Do not commit the value.
+   deploy time, `abuse_gate_429` and `abuse_gate_unavailable` log lines emit `ip_hash: null`,
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply docs/runbook.md:124]: Valid and fixed in replacement PR #596 (`4245e1ba`): `docs/ARCHITECTURE.md` now describes `IP_HASH_SECRET` for all abuse-gate log lines, matching the runbook’s `abuse_gate_429` and `abuse_gate_unavailable` behavior.

File: docs/runbook.md [hunk 5]
```diff
@@ -112,7 +112,7 @@ product.
    production **before** merging the worker change; adding a function ahead of its caller is safe.
 
 4. **Pre-deploy secret check (api-gateway Worker):** before merging a change that relies on
-   `IP_HASH_SECRET` (e.g. any change to IP-backstop logging), confirm the secret is already
+   `IP_HASH_SECRET` (e.g. any change to abuse-gate IP logging), confirm the secret is already
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply docs/runbook.md:115]: Valid and fixed in replacement PR #596 (`4245e1ba`): the pre-deploy check is now scoped specifically to abuse-gate logs that emit `ip_hash`.

File: docs/API.md [hunk 6]
```diff
@@ -375,23 +375,25 @@ This section covers **external progress API quotas only** (Worker + Durable Obje
...
+Authenticated responses include `X-RateLimit-Limit`, `X-RateLimit-Remaining`, and `X-RateLimit-Reset` (Unix seconds) for the daily quota, plus `Retry-After` on `429` responses. Pre-authentication abuse-gate `429` responses include only `Retry-After`, and fail-open responses (daily-quota service temporarily unavailable) omit the `X-RateLimit-*` headers. When a free-tier user exhausts a daily quota, the `429` body includes an upgrade link. Admins can inspect the top consumers via `GET /api/admin/api-usage`; usage is bucketed by UTC day, so the report covers the current and previous UTC day (the `since` field gives the exact starting day).
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply docs/API.md:396]: Valid and fixed in replacement PR #596 (`4245e1ba`): the API guarantee is now limited to responses for which the daily-quota service returned a quota decision.

**Comment by DysektAI** [comment]
Superseded by #596, which is based directly on the merged #591 commit and contains only the follow-up fixes.

All review feedback from this PR was reconciled there:

- CodeRabbit’s test-cleanup and lazy-payload suggestions are included in `4245e1ba`.
- Cubic’s API/runbook/architecture documentation findings are included in `4245e1ba`.
- Devin’s three informational observations were confirmed and explicitly disposed in their threads.
- Kilo produced no findings because its agent wrapper failed; the complete diff received a direct self-review and all other automated checks/reviews completed.

Closing this conflicted duplicate in favor of #596.
